### PR TITLE
Apple Music radio stations and improved URL import support

### PR DESCRIPTION
**Pull request by dmoo500**
## Summary

This PR adds Apple Music Artist Radio station playback support and improves how Apple Music share URLs are handled throughout Music Assistant.

> **Note:** This feature was developed by @dmoo500 with the assistance of GitHub Copilot (AI). The author is not deeply familiar with all parts of the codebase — please review thoroughly before merging.

---

### Apple Music Artist Radio stations via "Add URL"
Apple Music share URLs for (Artist Radio) stations can now be added using the "Add URL" dialog:
- Example: `https://music.apple.com/de/station/dead-sara-essentials/ra.331701075`

**Track-based Artist Radio stations** (e.g. "Dead Sara und ähnliche Künstler:innen", `ra.xxxxxxxxx`)
- Individual tracks fetched via `me/stations/next-tracks/{id}` API
- Auto-plays next track when current track ends (~5s gap)
- Skip button fetches a fresh track from the station
- Currently playing artist and track title shown on the player

### Improved Apple Music share URL routing
All Apple Music share URL types now resolve to the correct provider and media type:

| URL type | Example |
|---|---|
| Station | `…/station/…/ra.xxx` |
| Playlist | `…/playlist/…/pl.xxx` |
| Album | `…/album/…/12345` |
| Artist | `…/artist/…/12345` |
| Song | `…/song/…/12345` |

Previously only station URLs were handled; playlist/album/artist/song URLs fell through to the `builtin` provider and failed with ffprobe errors.

When a URL resolves to a different media type than the calling controller, it is now automatically delegated to the correct controller.

---

### Live broadcast stations do not work
**Apple Music live broadcast stations** (Apple Music 1, Apple Music Country, third-party stations like Radio Bob, etc.) **cannot be streamed**. All attempts via the `webPlayback` endpoint return error **3077** (`MZCommerce.ItemNotFoundForFuse`). The `me/stations/next-tracks` endpoint also returns an empty list for those stations.

The Apple Music app likely uses internal mechanisms to stream live stations that are not accessible through the public API. There is currently no known workaround. For those stations, using an alternative source (e.g. Radio Browser provider, or the station's own direct stream URL) is recommended.

### No automatic library sync for radio stations
`LIBRARY_RADIOS` / `get_library_radios()` is **not implemented**. Radio stations must be added manually via "Add URL". An earlier attempt at auto-syncing Apple Music stations was unreliable and has been dropped from this PR.

---

## Files changed

- `music_assistant/providers/apple_music/__init__.py` — Main provider: radio stream logic, URL handling, metadata helpers
- `music_assistant/helpers/uri.py` — Apple Music URL parsing for all share URL types
- `music_assistant/controllers/media/base.py` — URL routing to correct controller; human-readable name from URL slug
- `music_assistant/controllers/player_queues.py` — Continuous playback, skip button, and artist/title display for track-based radio

### CODE REVIEW

**Comment by dmoo500** [comment]
Could a maintainer please add the labels `new-feature` and `enhancement`? Thanks!

File: music_assistant/providers/apple_music/__init__.py [hunk 1]
```diff
@@ -692,6 +790,101 @@ async def set_favorite(self, prov_item_id: str, media_type: MediaType, favorite:
...
+                {"stationId": station_id, "stationHash": station_hash},
+                {"salableAdamId": numeric_id, "stationHash": station_hash},
+            ]
+        last_error: str = ""
+        for data in attempts:
+            async with self.mass.http_session.post(
+                playback_url, headers=self._get_decryption_headers(), json=data, ssl=True
+            ) as response:
+                response.raise_for_status()
+                content = await response.json(loads=json_loads)
```
[the remark this answers is not in the document]
Comment (dmoo500) [reply music_assistant/providers/apple_music/__init__.py]: Fixed in 68e31054 — wrapped the `post`/`raise_for_status`/`json` block in `except ClientError` and re-raise as `MediaNotFoundError`.

File: music_assistant/controllers/player_queues.py [hunk 2]
```diff
@@ -920,6 +920,16 @@ async def next(self, queue_id: str) -> None:
             return
         next_index = self._get_next_index(queue_id, idx, True)
         if next_index is None:
+            # For RADIO items there's no next index, but we can skip to the next
+            # station track by re-playing the same index (provider fetches a new track).
+            current_item = self.get_item(queue_id, idx)
```
[the remark this answers is not in the document]
Comment (dmoo500) [reply music_assistant/controllers/player_queues.py:925]: Fixed in 68e31054 — added `self._transitioning_players.discard(queue_id)` in the `else` branch before returning.

File: music_assistant/helpers/uri.py [hunk 3]
```diff
@@ -46,6 +46,29 @@ async def parse_uri(uri: str, validate_id: bool = False) -> tuple[MediaType, str
...
+        elif uri.startswith("https://music.apple.com/"):
+            # Apple Music share URL
+            # https://music.apple.com/{storefront}/{type}/{slug}/{id}
+            _apple_type_map = {
+                "station": MediaType.RADIO,
+                "playlist": MediaType.PLAYLIST,
+                "album": MediaType.ALBUM,
+                "artist": MediaType.ARTIST,
+                "song": MediaType.TRACK,
+            }
```
[the remark this answers is not in the document]
Comment (dmoo500) [reply music_assistant/helpers/uri.py:58]: Added `test_apple_music_uri_parsing()` in `tests/core/test_helpers.py` covering all 5 URL types (station, playlist, album, artist, song) plus trailing-slash and query-string edge cases (commit 68e31054).

**Comment by MarvinSchenkel** [comment]
I'm sorry but we cannot accept this PR. This PR contains a lot of changes to code **outside** of the Apple Music provider. This should not be required when adding functionality to a music provider, unless there is a very good reason to do so. Changes should be isolated to files inside the apple music provider.

Please re-evaluate your code and get rid of changes to files outside of the Apple music provider unless you think it is absolutely necessary. Marking this PR as draft, please click 'ready for review' again when you want us to have another look.

**Comment by dmoo500** [comment]
Thank you for the feedback. I've reviewed all external changes and reduced them to the absolute minimum needed.

**Current diff outside the Apple Music provider:**

### `helpers/uri.py` (+23 lines)
Adding Apple Music URL parsing here follows the established pattern — Spotify, Tidal and Qobuz share URLs are all handled in this same file. It is not possible to place this inside the provider itself, as `parse_uri` has no access to a provider instance.

### `controllers/media/base.py` (+4 lines)
This is the minimal controller-layer change required. The `music/radios/get_radio` API endpoint calls `base.get(item_id, provider)` directly, and the frontend passes the full share URL as `item_id`. Without this 4-line hook, the URL cannot be resolved to `apple_music` and falls through to the `builtin` provider, which tries to ffprobe the Apple Music URL and fails with `InvalidDataError`.

### `controllers/player_queues.py` (+48 lines)
These changes are intentionally **provider-agnostic** — they improve radio playback for **any** provider that uses track-based stations (continuous playback when a track ends, skip button support, `stream_metadata` display for artist/title). Moving this logic into the Apple Music provider is not possible because providers have no access to the queue controller.

### `tests/core/test_helpers.py` (+51 lines)
Unit tests for the `uri.py` changes. They belong here alongside the existing `test_uri_parsing` tests.

I'm happy to discuss further if you'd like a different approach for any of these.

**Comment by OzGav** [comment]
I would say your changes to player_queues.py should be in its own PR

**Comment by dmoo500** [comment]
Following review feedback, I split this work into two separate PRs:

1. #3433 Apple Music provider + URL import support (without queue controller changes)
2. #3432 provider-agnostic queue behavior improvements in player_queues.py

If preferred, we can continue review on those split PRs instead of this combined one.
